### DISCUSSION TITLE
Adding Macros to a SyGuS Grammar

### DISCUSSION THREAD

**Discussion by NeumannDirk** [cvc5/cvc5#10509] [Q&A]
Hi,

I am trying to work with SyGuS and cvc5 in the context of arrays. For this, I am using [sequences](https://microsoft.github.io/z3guide/docs/theories/Sequences/), but I am missing a simple `seq.update`. So my idea was to add it as a macro myself. How can I achieve his without having to redefine the rest of the grammar?

In [The SyGuS Language Standard ](https://arxiv.org/pdf/2312.06001.pdf), page 25, example 4, there is a macro used, named _x_plus_one_, in the synthesis of the function `f`. In this case, it is fairly easy since the grammar is small anyway. So I want to give a bigger example:

```
(set-logic SLIA)

;I want to edit multiple values and therefore I wrap the in my own tuple
(declare-datatype Tuple ((mkTuple (getField0 Int) (getField1 (Seq Int)))))

,defining pre and pos conditions
(declare-var i_preCon Int)
(declare-var A_preCon Int)
(declare-var i_postCon Int)
(declare-var A_postCon Int)

(define-fun preCondition ((i_preCon Int) (A_preCon (Seq Int))) Bool 
    (= 5 (seq.nth A_preCon i_preCon))
)

(define-fun postCondition ((i_postCon Int) (A_postCon (Seq Int))) Bool 
    (= 6 (seq.nth A_postCon i_postCon))
)

;defining my macro
(define-fun seqUpdate ((s (Seq Int)) (i Int) (val Int)) (Seq Int)
    (seq.++ (seq.extract s 0 i) (seq.unit val) (seq.extract s (+ i 1) (- (seq.len s) i 1)))
)

(synth-fun targetFunction ((i Int) (A (Seq Int))) Tuple)

;constraints
(declare-var i_in Int)
(declare-var A_in (Seq Int))
(declare-var i_out Int)
(declare-var A_out (Seq Int))

(constraint (forall ((i_in Int) (i_out Int) (A_in (Seq Int)) (A_out (Seq Int))) (=>
	(and
		(preCondition i_in A_in)
		(= i_out (getField0 (targetFunction i_in A_in)))
		(= A_out (getField1 (targetFunction i_in A_in)))
	)
	(and
		(postCondition i_out A_out)
		(= i_in i_out)
	)
)))
(check-synth)
```

I am generating my sygus problems programmatically. The synthesized program shall represent an update for each variable I put in. So, the issue is, that cvc5 does not want to use `seqUpdate` in this version. So what I did so far is:

1. run `cvc5 --output sygus-grammar myProblem.sy`
2. take the grammar from the output and add this grammar together with my own macro to the sygus file (by hand)
This results in a grammar like this:
```
(synth-fun targetFunction ((i Int) (A (Seq Int))) Tuple
((A_Tuple Tuple) (A_Int Int) (A_Seq_336 (Seq Int)) (A_Bool Bool))
(
    (A_Tuple Tuple ((mkTuple A_Int A_Seq_336) (ite A_Bool A_Tuple A_Tuple)))
    (A_Int Int (i 5 6 0 1 (+ A_Int A_Int) (- A_Int A_Int) (str.len A_Seq_336) (getField0 A_Tuple) (ite A_Bool A_Int A_Int)))
;scroll right
    (A_Seq_336 (Seq Int) (A (as seq.empty (Seq Int)) (str.++ A_Seq_336 A_Seq_336) (seq.unit A_Int) (getField1 A_Tuple) (ite A_Bool A_Seq_336 A_Seq_336) **(seqUpdate A_Seq_336 A_Int A_Int)**))
    (A_Bool Bool (true false (= A_Int A_Int) (<= A_Int A_Int) (= A_Seq_336 A_Seq_336) (= A_Tuple A_Tuple) ((_ is mkTuple) A_Tuple) (not A_Bool) (and A_Bool A_Bool) (or A_Bool A_Bool)))
))
```
3. run `cvc5 --output sygus myProblem.sy`
Now I can see that `seqUpdate` is used in different candidates. 

Having a first call to get the grammar and then copying it to the problem with some changes to make the second call seems quite tedious. Is there a simpler way to just list the macro instead of having the two calls? Any hint is highly appreciated.

Thank you

**Comment by NeumannDirk** [comment]
Hello,

I would like to ask if there is an answer to my question. A solution to the problem described would be a great help in speeding up the work on my own research. I would be very grateful for a prompt reply.

Best regards
Dirk

**Comment by ajreynol** [comment]
Sorry for the delay in reply, we currently don't have support for providing partial information for a grammar.

We could potentially support this by supplying a list of ground terms to consider as terminal symbols of the grammar. This would include function symbols defined by macros.

I will keep you posted if we add this to our API.
